File: typescript/packages/bench/src/support/abstractExporter.ts
/** Base shared by all 8 exporters. */

import type { ExporterContract } from "../contracts/exporterContract";
import type { FormatterContract } from "../contracts/formatterContract";
import type { ReportRow } from "../contracts/reportContract";

export abstract class AbstractExporter implements ExporterContract, FormatterContract {
    protected constructor(
        readonly extension: string,
        readonly mime: string,
    ) {}

    abstract export(rows: readonly ReportRow[]): string;

    filename(slug: string): string {
        return `${slug}.${this.extension}`;
    }

    formatCents(cents: number): string {
        const sign = cents < 0 ? "-" : "";
        const abs = Math.abs(cents);

        return `${sign}${Math.floor(abs / 100)}.${String(abs % 100).padStart(2, "0")}`;
    }

    formatPercent(ratio: number): string {
        return `${(ratio * 100).toFixed(1)}%`;
    }

    formatCount(count: number): string {
        return String(count);
    }
}

File: python/atelier_bench/support/abstract_exporter.py
"""Base exporter: format_cents / format_percent / format_count arrive from FormatterContract through the MRO."""

from __future__ import annotations

from ..contracts.exporter_contract import ExporterContract, filename_for
from ..contracts.formatter_contract import FormatterContract


class AbstractExporter(ExporterContract, FormatterContract):
    def __init__(self, extension: str, mime: str) -> None:
        self.extension = extension
        self.mime = mime

    def filename(self, slug: str) -> str:
        return filename_for(slug, self.extension)
